### PR TITLE
feat: New `Config` options for numeric formatting: digit grouping and thousands/decimal separator

### PR DESCRIPTION
**Pull request by alexander-beedie**
Ref: https://github.com/pola-rs/polars/issues/8157, https://github.com/pola-rs/polars/issues/7006.

## New options

Adds new `Config` options to support digit-grouping and locale-specific thousands/decimal separators for int/float/decimal values on display (eg: in table repr):

* ~`set_digit_group_size`~
* ~`set_digit_group_separator`~

(Update; dropped the two settings above in favour of `thousands_separator`, below).

* `set_thousands_separator`
* `set_decimal_separator`

## Examples

Add thousands separator; note that this is most (visually) appealing in conjunction with right-aligned columns and (if float) with a fixed decimal precision:
```python
import polars as pl

df = pl.DataFrame({
    "x": ["aa", "bb", "cc"],
    "y": [9876543.21, -2345.9876, 101010.],
    "z": [123456789, 9876543, -101010],
})
with pl.Config( thousands_separator=True ):
    print( df )
    
# shape: (3, 3)
# ┌─────┬──────────────┬─────────────┐
# │ x   ┆ y            ┆ z           │
# │ --- ┆ ---          ┆ ---         │
# │ str ┆ f64          ┆ i64         │
# ╞═════╪══════════════╪═════════════╡
# │ aa  ┆ 9,876,543.21 ┆ 123,456,789 │
# │ bb  ┆ -2,345.9876  ┆ 9,876,543   │
# │ cc  ┆ 101,010.0    ┆ -101,010    │
# └─────┴──────────────┴─────────────┘
```
The same example, but with the other two options enabled - looks _much_ better 👀 
```python
with pl.Config(
    tbl_cell_numeric_alignment = "RIGHT",
    thousands_separator = True,
    float_precision = 3,
):
    print( df )

# shape: (3, 3)
# ┌─────┬───────────────┬─────────────┐
# │ x   ┆             y ┆           z │
# │ --- ┆           --- ┆         --- │
# │ str ┆           f64 ┆         i64 │
# ╞═════╪═══════════════╪═════════════╡
# │ aa  ┆ 9,876,543.210 ┆ 123,456,789 │
# │ bb  ┆    -2,345.988 ┆   9,876,543 │
# │ cc  ┆   101,010.000 ┆    -101,010 │
# └─────┴───────────────┴─────────────┘
```
Set custom thousands/decimal separators:
_(Covers large parts of Europe / Latin America, which invert usage of the "." and "," chars)_
```python
with pl.Config(
    tbl_cell_numeric_alignment = "RIGHT",
    thousands_separator = ".",
    decimal_separator = ",",
    float_precision = 3,
):
    print( df )

# shape: (3, 3)
# ┌─────┬───────────────┬─────────────┐
# │ x   ┆             y ┆           z │
# │ --- ┆           --- ┆         --- │
# │ str ┆           f64 ┆         i64 │
# ╞═════╪═══════════════╪═════════════╡
# │ aa  ┆ 9.876.543,210 ┆ 123.456.789 │
# │ bb  ┆    -2.345,988 ┆   9.876.543 │
# │ cc  ┆   101.010,000 ┆    -101.010 │
# └─────┴───────────────┴─────────────┘
```

## Note

This is not a complete solution covering more complex locale-specific formats, such as Indian (which _mixes_ group lengths in a single number, eg: `12,34,567`) or the German DIN1333 standard (which uses spaces, but _only_ once the integer part is >= 5 digits, eg: `1234` vs `12 345`). 

Perhaps we can go there later, but real support for all major locale conventions is not a trivial endeavour; probably a better move to take on an additional dependency from a dedicated crate[^1] _if_ we ever decide that we want that.

## Also

* Fixes the `tbl_cell_numeric_alignment` option so that it accounts for decimal dtypes.
* Adds a `set_trim_decimal_zeros` option to make that behaviour configurable.
* Added some missing `Config` options to the docs.
* Various improvements to `pl.from_repr`.

[^1]: Such as `num_format` (https://docs.rs/num-format/latest/num_format/)

### CODE REVIEW

**Comment by alexander-beedie** [comment]
Will clean up the clippy & docs build issues shortly 👌

**Comment by alexander-beedie** [comment]
Looks like some Config state is bleeding through in the doctests; will find/fix 🤔

**Comment by alexander-beedie** [comment]
Rebased, following https://github.com/pola-rs/polars/pull/12096.

File: py-polars/src/lib.rs [hunk 1]
```diff
@@ -225,13 +225,31 @@ fn polars(py: Python, m: &PyModule) -> PyResult<()> {
         functions::string_cache::using_string_cache
     ))
     .unwrap();
-    m.add_wrapped(wrap_pyfunction!(functions::meta::set_float_fmt))
+
+    // Numeric formatting
+    m.add_wrapped(wrap_pyfunction!(functions::meta::get_digit_group_separator))
+        .unwrap();
```
Comment (stinodego) [inline py-polars/src/lib.rs:231]: Just for my understanding: why are some of  the Config options implemented as functions in the Rust bindings, while others go through environment variables?

File: py-polars/polars/utils/various.py [hunk 2]
```diff
@@ -339,7 +342,35 @@ def str_duration_(td: str | None) -> int | None:
...
+                    # check for empty string and/or integer format
+                    pl.when(F.col(c).str.contains(r"^[+-]?\d*$"))
+                    .then(pl.when(F.col(c).str.len_bytes() > 0).then(F.col(c)))
+                    # check for scientific notation
+                    .when(F.col(c).str.contains("[eE]"))
+                    .then(F.col(c).str.replace(r"[^eE\d]", "."))
+                    .otherwise(
+                        # recombine sanitised integer/fractional components
+                        pl.concat_str(
+                            integer_part.str.replace_all(r"[^\d+-]", ""),
```
Comment (stinodego) [inline py-polars/polars/utils/various.py:367]: The Lord of RegEx strikes again 😄

File: py-polars/polars/series/string.py [hunk 3]
```diff
@@ -314,9 +314,9 @@ def to_decimal(
             3420.13
             120134.19
             3212.98
-            12.9
+            12.90
             143.09
-            143.9
+            143.90
```
Comment (stinodego) [inline py-polars/polars/series/string.py:319]: Nice that this prints a consistent number of decimals now!

File: py-polars/docs/source/reference/config.rst [hunk 4]
```diff
@@ -11,11 +11,18 @@ Config options
...
+    Config.set_decimal_separator
+    Config.set_digit_group_separator
+    Config.set_digit_group_size
+    Config.set_float_precision
     Config.set_fmt_float
     Config.set_fmt_str_lengths
     Config.set_fmt_table_cell_list_len
     Config.set_streaming_chunk_size
     Config.set_tbl_cell_alignment
+    Config.set_tbl_cell_alignment
```
Comment (stinodego) [inline py-polars/docs/source/reference/config.rst:24]: Accidentally duplicated line.

```suggestion
```

File: crates/polars-core/src/datatypes/dtype.rs [hunk 5]
```diff
@@ -166,6 +166,16 @@ impl DataType {
         self.is_float() || self.is_integer()
     }
 
+    /// Check if this [`DataType`] is a Decimal type (of any scale/precision).
+    pub fn is_decimal(&self) -> bool {
+        #[allow(clippy::match_like_matches_macro)]
```
Comment (stinodego) [inline crates/polars-core/src/datatypes/dtype.rs:171]: Why not just use the `matches!` macro here?

File: py-polars/polars/config.py [hunk 6]
```diff
@@ -20,14 +20,38 @@ def _get_float_precision() -> int:
...
         get_float_precision as _get_float_precision,
+        set_float_precision as _set_float_precision,
+        set_digit_group_separator as _set_digit_group_separator,
+        get_digit_group_separator as _get_digit_group_separator,
+        set_digit_group_size as _set_digit_group_size,
+        get_digit_group_size as _get_digit_group_size,
+        set_decimal_separator as _set_decimal_separator,
+        get_decimal_separator as _get_decimal_separator,
+        get_trim_decimal_zeros as _get_trim_decimal_zeros,
+        set_trim_decimal_zeros as _set_trim_decimal_zeros,
```
Comment (stinodego) [inline py-polars/polars/config.py:53]: Better to use `import polars.polars as plr` and then use `plr.set_float_fmt` etc. I have it like that in a few places, apparently I hadn't updated this file.

File: py-polars/polars/config.py [hunk 7]
```diff
@@ -20,14 +20,38 @@ def _get_float_precision() -> int:
...
+def _get_digit_group_size() -> int:
+    return 3
+
+
+def _get_decimal_separator() -> str:
+    return "."
+
+
+def _get_trim_decimal_zeros() -> bool:
+    return False
```
Comment (stinodego) [inline py-polars/polars/config.py:36]: I don't think these should be here?

**Review by stinodego — CHANGES_REQUESTED** [review]
Really nice stuff. I left a few minor comments!

File: py-polars/src/lib.rs (hunk 1 above)
Comment (alexander-beedie) [reply py-polars/src/lib.rs:231]: Latency, I believe; want to avoid repeated env var lookup in a loop.

**Comment by mcrumiller** [comment]
Are there parts of the world that group digits by a value other than 3? I'm questioning the need for `digit_group_size` as opposed to binary `group_digits`.

**Comment by mcrumiller** [comment]
I would recommend in that case to use the `thousands_separator` keyword instead of `group_digits`, as the former is a bit more familiar.

File: py-polars/polars/config.py (hunk 7 above)
Comment (alexander-beedie) [reply py-polars/polars/config.py]: They don't now that I've integrated the `plr` suggestion  :)

File: py-polars/polars/config.py (hunk 6 above)
Comment (alexander-beedie) [reply py-polars/polars/config.py]: Done; good tidy-up 👍

File: crates/polars-core/src/datatypes/dtype.rs (hunk 5 above)
Comment (alexander-beedie) [reply crates/polars-core/src/datatypes/dtype.rs:171]: Didn't play well with the requirement to use `feature = "dtype-decimal"` (unless some syntax nicety is available that I'm unaware of?), but I'm not sure how that `#[allow(clippy::match_like_matches_macro)]` ended up there - removing!

**Comment by alexander-beedie** [comment]
> I would suggest getting rid of the parameter and just using a fixed grouping of 3 digits.

Will take care of that one tomorrow; just finished off the other suggestions 👌

**Review by stinodego — APPROVED** [review]
This looks good to me! Maybe someone wants to double-check the Rust side of things.

And would be good for @ritchie46 to weigh in as it's a change to the public API.

**Comment by alexander-beedie** [comment]
(Updated the examples shown in the description to use the latest `thousands_separator` config option).

**Comment by alexander-beedie** [comment]
Rebased.

**Review by ritchie46 — APPROVED** [review]

**Pull request merged**